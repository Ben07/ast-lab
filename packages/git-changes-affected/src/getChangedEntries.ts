import { parse, ParserOptions } from '@babel/parser';
import { File, SourceLocation } from '@babel/types';
import { extractStats, getPrivateMemberNames } from 'es-stats';
import _debug from 'debug';
import getAbsolutePath from './getAbsolutePath';
import { Entry, MemberRef, MemberRelation } from 'ast-lab-types';
import { Change } from './types';


const debug = _debug('git-changes-affected:entries');

type Declare = MemberRef & { loc: SourceLocation };
/** @todo improve precision of private declaration location */
function locatePrivateDeclares(ast: File): Declare[] {
  return ast.program.body.reduce((locs, d) => {
    const refs = getPrivateMemberNames(d);
    const { loc } = d;
    if (!refs || !refs.length || !loc) { return locs; }
    return locs.concat(
      refs.map(ref => ({
        ...ref,
        loc
      }))
    );
  }, [] as Declare[]);
}

export default function getChangedEntries(changes: Change[], parserOptions?: ParserOptions | null) {
  return changes.reduce((res, { file, content, changed }) => {
    if (!content) {
      return res;
    }
    const filePath = getAbsolutePath(file);
    let ast;
    try { 
      ast = (parse(content, { 
        ...(parserOptions || {}),
        sourceType: 'module'
      }) as File);
    } catch(e) {
      console.warn(`@bable/parser parsing ${filePath} failed! (${e.message}) Parser options: ${JSON.stringify(parserOptions)}`);
      console.warn('File content >>> \n', content);
      return res;
    }
    const stats = extractStats(ast);
    debug(`${file} stats: ${JSON.stringify(stats, null, 2)}`);
    const declareLoc = [
      ...locatePrivateDeclares(ast),
      ...(stats.exports.members as Declare[]),
      ...(stats.imports as Declare[])
    ];
    debug(`${file} declareLoc: ${JSON.stringify(declareLoc, null, 2)}`);
    const exported = new Set(stats.exports.members.map(({ alias }) => alias));
    const affectExports = {} as MemberRelation;
    exported.forEach(name => {
      const expanded = new Set(stats.relations[name] || []);
      expanded.forEach(m => {
        if (m === name) { return; }
        affectExports[m] = [
          ...(affectExports[m] || []),
          name
        ];
        if (stats.relations[m]) {
          stats.relations[m].forEach(r => expanded.add(r));
        }
      });
    });

    let iDeclare = 0;
    const changedExports = changed.reduce(
      (ex, { start: startLine, end: endLine }) => {
        while (iDeclare < declareLoc.length) {
          const { loc, alias } = declareLoc[iDeclare];
          iDeclare++;
          if (!loc) { continue; }
          const { start, end } = loc;
          if (endLine < start.line) { 
            return ex;
          }
          if (startLine <= end.line) {
            if (exported.has(alias)) {
              ex.push(alias);
            } else if (affectExports[alias]) {
              ex = ex.concat(affectExports[alias]);
            }
          }
        }
        return ex;
      }, 
      [] as string[]
    );
    return res.concat(
      Array.from(new Set(changedExports)).map(name => ({
        source: filePath,
        name
      }))
    );
  }, [] as Entry[]);
}