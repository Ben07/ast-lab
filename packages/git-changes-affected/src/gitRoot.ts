import exec from './exec';

export default function gitRoot() {
  return exec('git rev-parse --show-toplevel');
}