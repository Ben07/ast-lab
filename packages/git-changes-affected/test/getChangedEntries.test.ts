import path from 'path';
import getChangedEntries from '../src/getChangedEntries';
import gitRoot from '../src/gitRoot';
import { Entry } from 'ast-lab-types';

const sampleFile = `import a from 'SampleModule';

export const b = 1;

export function func() {
 let d = b + 1;
 return a(d);
}

export default func;
`;

const ROOT = gitRoot();
function relativeEntries(entries: Entry[]) {
  return entries.map(({ name, source }) => {
    return {
      name,
      source: path.relative(ROOT, source),
    };
  });
}

describe('getChangedEntries', () => {
  test('import could be counted as entry', () => {
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 1, end: 1 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
  });
  test('single line', () => {
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 3, end: 3 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 5, end: 5 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 8, end: 8 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
  });
  test('within 1 export', () => {
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 5, end: 6 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
  });
  test('cross exports', () => {
    expect(
      relativeEntries(
        getChangedEntries(
          [
            {
              file: 'sampleFile.js',
              content: sampleFile,
              changed: [{ start: 2, end: 6 }],
            },
          ],
          {
            sourceType: 'module',
          }
        )
      )
    ).toMatchSnapshot();
  });
});
