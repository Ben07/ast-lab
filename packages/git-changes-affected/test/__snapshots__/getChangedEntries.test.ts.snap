// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChangedEntries cross exports 1`] = `
Array [
  Object {
    "name": "b",
    "source": "sampleFile.js",
  },
  Object {
    "name": "func",
    "source": "sampleFile.js",
  },
  Object {
    "name": "default",
    "source": "sampleFile.js",
  },
]
`;

exports[`getChangedEntries import should not be counted as entry 1`] = `Array []`;

exports[`getChangedEntries single line 1`] = `
Array [
  Object {
    "name": "b",
    "source": "sampleFile.js",
  },
  Object {
    "name": "func",
    "source": "sampleFile.js",
  },
  Object {
    "name": "default",
    "source": "sampleFile.js",
  },
]
`;

exports[`getChangedEntries single line 2`] = `
Array [
  Object {
    "name": "func",
    "source": "sampleFile.js",
  },
  Object {
    "name": "default",
    "source": "sampleFile.js",
  },
]
`;

exports[`getChangedEntries single line 3`] = `
Array [
  Object {
    "name": "func",
    "source": "sampleFile.js",
  },
  Object {
    "name": "default",
    "source": "sampleFile.js",
  },
]
`;

exports[`getChangedEntries within 1 export 1`] = `
Array [
  Object {
    "name": "func",
    "source": "sampleFile.js",
  },
  Object {
    "name": "default",
    "source": "sampleFile.js",
  },
]
`;
